File: booking_service/src/core/service/accommodation_service.go
package service

import (
	"booking_service/core/domain/dto/request"
	"booking_service/core/domain/entity"
	"booking_service/core/usecase"
	"context"
)

type IAccommodationService interface {
	CreateAccommodation(ctx context.Context, accommodation *entity.AccommodationEntity) (*entity.AccommodationEntity, error)
	GetAccommodationByID(ctx context.Context, ID int64) (*entity.AccommodationEntity, error)
	UpdateAccommodation(ctx context.Context, accommodation *entity.AccommodationEntity) (*entity.AccommodationEntity, error)
	DeleteAccommodation(ctx context.Context, ID int64) error
	AddUnitToAccommodation(ctx context.Context, req *request.AddUnitToAccommodationDto) (*entity.UnitEntity, error)
}

type AccommodationService struct {
	createAccUseCase usecase.ICreateAccommodationUseCase
	getAccUseCase    usecase.IGetAccommodationUseCase
	deleteAccUseCase usecase.IDeleteAccommodationUseCase
	updateAccUseCase usecase.IUpdateAccommodationUseCase
}

func (a AccommodationService) AddUnitToAccommodation(ctx context.Context, req *request.AddUnitToAccommodationDto) (*entity.UnitEntity, error) {
	return a.updateAccUseCase.AddUnitToAccommodation(ctx, req)
}

func (a AccommodationService) CreateAccommodation(ctx context.Context, accommodation *entity.AccommodationEntity) (*entity.AccommodationEntity, error) {
	return a.createAccUseCase.CreateAccommodation(ctx, accommodation)
}

func (a AccommodationService) GetAccommodationByID(ctx context.Context, ID int64) (*entity.AccommodationEntity, error) {
	return a.getAccUseCase.GetAccommodationByID(ctx, ID)
}

func (a AccommodationService) UpdateAccommodation(ctx context.Context, accommodation *entity.AccommodationEntity) (*entity.AccommodationEntity, error) {
	return a.updateAccUseCase.UpdateAccommodationByID(ctx, accommodation)
}

func (a AccommodationService) DeleteAccommodation(ctx context.Context, ID int64) error {
	return a.deleteAccUseCase.DeleteAccommodationByID(ctx, ID)
}

func NewAccommodationService(createAccUseCase usecase.ICreateAccommodationUseCase,
	getAccUseCase usecase.IGetAccommodationUseCase,
	deleteAccUseCase usecase.IDeleteAccommodationUseCase,
	updateAccUseCase usecase.IUpdateAccommodationUseCase) IAccommodationService {
	return &AccommodationService{
		createAccUseCase: createAccUseCase,
		getAccUseCase:    getAccUseCase,
		deleteAccUseCase: deleteAccUseCase,
		updateAccUseCase: updateAccUseCase,
	}
}

File: search_service/src/main/java/huy/project/search_service/core/service/AccommodationService.java
package huy.project.search_service.core.service;

import huy.project.search_service.core.domain.dto.request.AccommodationParams;
import huy.project.search_service.core.domain.dto.response.AccommodationThumbnail;
import huy.project.search_service.core.domain.dto.response.PageInfo;
import huy.project.search_service.core.domain.entity.AccommodationEntity;
import huy.project.search_service.core.domain.entity.UnitEntity;
import huy.project.search_service.core.usecase.*;
import lombok.AccessLevel;
import lombok.RequiredArgsConstructor;
import lombok.experimental.FieldDefaults;
import org.springframework.data.util.Pair;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@RequiredArgsConstructor
@FieldDefaults(level = AccessLevel.PRIVATE, makeFinal = true)
public class AccommodationService implements IAccommodationService{
    CreateAccommodationUseCase createAccommodationUseCase;
    GetAccommodationUseCase getAccommodationUseCase;
    DeleteAccommodationUseCase deleteAccommodationUseCase;
    AddUnitUseCase addUnitUseCase;
    DeleteUnitUseCase deleteUnitUseCase;
    UpdateAccommodationUseCase updateAccommodationUseCase;

    @Override
    public AccommodationEntity createAccommodation(AccommodationEntity accommodation) {
        return createAccommodationUseCase.createAccommodation(accommodation);
    }

    @Override
    public AccommodationEntity getAccById(Long id) {
        return getAccommodationUseCase.getAccById(id);
    }

    @Override
    public void deleteAccommodation(Long id) {
        deleteAccommodationUseCase.deleteAccommodation(id);
    }

    @Override
    public void addUnitToAccommodation(Long accId, UnitEntity unit) {
        addUnitUseCase.addUnitToAccommodation(accId, unit);
    }

    @Override
    public Pair<PageInfo, List<AccommodationEntity>> getAccommodations(AccommodationParams params) {
        return getAccommodationUseCase.getAccommodations(params);
    }

    @Override
    public Pair<PageInfo, List<AccommodationThumbnail>> getAccommodationsThumbnail(AccommodationParams params) {
        return getAccommodationUseCase.getAccommodationsThumbnail(params);
    }

    @Override
    public void deleteUnit(Long accId, Long unitId) {
        deleteUnitUseCase.deleteUnit(accId, unitId);
    }

    @Override
    public void updateAccommodation(AccommodationEntity accommodation) {
        updateAccommodationUseCase.updateAccommodation(accommodation);
    }
}
